packages/streamr-eventemitter/include/streamr-eventemitter/EventEmitter.hpp

```diff
@@ -0,0 +1,190 @@
+#ifndef STREAMR_EVENTEMITTER_HPP
+#define STREAMR_EVENTEMITTER_HPP
+
+#include <cstddef>
+#include <functional>
+#include <list>
+#include <mutex>
+#include <optional>
+#include <tuple>
+#include <type_traits>
+
+namespace streamr::eventemitter {
+
+template <typename... HandlerArgumentTypes>
+
+// Event class that all events must inherit from.
+// Usage example:
+// struct MyEvent : Event<int, std::string> {};
+struct Event {
+    class Handler {
+    public:
+        using HandlerFunction = std::function<void(HandlerArgumentTypes...)>;
+
+    private:
+        HandlerFunction handlerFunction;
+        size_t id;
+        bool once;
+
+    public:
+        explicit Handler(
+            HandlerFunction callback, size_t handlerId, bool once = false)
+            : handlerFunction(callback), id(handlerId), once(once) {}
+
+        void operator()(HandlerArgumentTypes... args) {
+            handlerFunction(std::forward<HandlerArgumentTypes>(args)...);
+        }
+
+        bool operator==(const Handler& other) const { return id == other.id; }
+
+        [[nodiscard]] bool isOnce() const { return once; }
+
+        [[nodiscard]] size_t getId() const { return id; }
+    };
+};
+
+// Each event type gets generated its own EventEmitterImpl
+template <typename EmitterEventType>
+class EventEmitterImpl {
+private:
+    std::list<typename EmitterEventType::Handler> eventHandlers;
+    std::mutex mutex;
+
+    // Immutable reference to a Handler.
+    // This is needed because there is no way of checking the validity
+    // of iterators returned by std::list, and removing a handler twice would
+    // crash the program if iterators were used.
+    class HandlerReference {
+    private:
+        size_t id;
+        explicit HandlerReference(size_t id) : id(id) {}
+
+    public:
+        static HandlerReference create() {
+            // Use a "magic static"
+            // https://blog.mbedded.ninja/programming/languages/c-plus-plus/magic-statics/
+            static size_t counter = 1;
+            return HandlerReference(counter++);
+        }
+        static HandlerReference createNonExistent() {
+            return HandlerReference(0);
+        }
+        [[nodiscard]] size_t getId() const { return id; }
+    };
+
+public:
+    template <typename EventType, typename CallbackType>
+        requires(
+            std::is_same<EventType, EmitterEventType>::value &&
+            std::is_assignable<
+                typename EmitterEventType::Handler::HandlerFunction,
```

Comment:
Why there is: typename EmitterEventType here. I thought that it is possible to have it only inside template. How is this even possible?